common/packet.go

```diff
@@ -0,0 +1,52 @@
+// Package common provides common functions for Geneva.
+package common
+
+import (
+	"encoding/binary"
+
+	"github.com/google/gopacket/layers"
+)
+
+// UpdateTCPChecksum updates the TCP checksum field and the raw bytes for a gopacket TCP layer.
+func UpdateTCPChecksum(tcp *layers.TCP) {
+	// the ComputeChecksum method requires the checksum bytes in the raw packet to be zeroed out.
+	tcp.Contents[16] = 0
+	tcp.Contents[17] = 0
+
+	chksum, _ := tcp.ComputeChecksum()
+
+	tcp.Checksum = chksum
+	binary.BigEndian.PutUint16(tcp.Contents[16:18], chksum)
+}
+
+// UpdateIPv4Checksum updates the IPv4 checksum field and the raw bytes for a gopacket IPv4 layer.
+func UpdateIPv4Checksum(ip *layers.IPv4) {
+	chksum := CalculateIPv4Checksum(ip.Contents)
+	ip.Checksum = chksum
+	binary.BigEndian.PutUint16(ip.Contents[10:12], chksum)
+}
+
+// CalculateIPv4Checksum calculates the IPv4 checksum for the given bytes.
+// copied from gopacket/layers/ip4.go because they didn't export one. for whatever some reason..
+func CalculateIPv4Checksum(bytes []byte) uint16 {
+	buf := make([]byte, len(bytes), 60)
+	copy(buf, bytes)
```

Comment:
If `len(bytes)` is ever larger than 60, this will panic. That said, I'll mention again that the fewer allocations and memory copies we can do in a critical packet-processing pipeline, the better. IIUC you should be able to clear the checksum bytes in the original packet slice and not have to worry about using this buffer.

Reply:
> If len(bytes) is ever larger than 60, this will panic.

Interesting, I thought it would automatically set `len` to `cap` if it was larger. 

> That said, I'll mention again that the fewer allocations and memory copies we can do in a critical packet-processing pipeline, the better. IIUC you should be able to clear the checksum bytes in the original packet slice and not have to worry about using this buffer.

The compiler will actually push the slice to the stack instead of allocating memory the size of the variable is small enough  (65535 bytes I think) and it doesn't escape the scope. Even when using `make`. I'm not sure which version implemented that change, it's pretty cool though. Either way, it isn't really necessary to copy the data. I'll remove it.

Reply:
Oh, that's pretty cool. I'll have to read up on that!